### ISSUE TITLE
QPF linearization returns exit code 3 on a valid PDF

### ISSUE THREAD

**Issue by foxmadr** [qpdf/qpdf#277]
Run on attached [h1.pdf](https://github.com/qpdf/qpdf/files/2772271/h1.pdf):
$ qpdf --linearize h1.pdf h2.pdf
or
$ qpdf --check

There is a warning:
WARNING: h1.pdf: reported number of objects (18) inconsistent with actual number of objects (17)

The PDF is valid, it was checked by some online tools. It causes exit code 3 so it means a failure.

The warning should not be there or exit code should be 0.

**Comment by jberkenbilt** [comment]
Thanks for the report. qpdf is behaving as documented. An exit code of 3 means that there were warnings, indicating that there were ways in which the file was non-compliant with the spec but that qpdf was able to recover. Virtually all PDF readers will ignore errors in the linearization tables, and most validators don't check them, but if you were writing PDFs, you'd want to know. If you are using qpdf to check PDFs, you should not consider an exit code of 3 to indicate a serious error in most cases. If you don't want to see the warning, you can use the `--no-warn` flag, but qpdf will still exit with an exit code of 3 if it encountered warnings.

I appreciate your taking the time to report this. I'm going to close this since it's working as designed, but feel free to reopen or comment further if needed. Thanks.

**Comment by foxmadr** [comment]
Thank you very much for explanation. I thought that the attached PDF was compliant. It just overrides ref object "3 0". It's presented twice (in both reference tables), but it was counted just once by qpdf.

**Comment by jberkenbilt** [comment]
Okay, thanks for the clarification. I'll leave this open until I have a chance to determine whether qpdf is incorrectly reporting a problem. I wrote the linearization code a very long time ago (more than 10 years), and most of the details have long since evaporated from my brain.

**Comment by foxmadr** [comment]
Thanks. I think that it's not problem of a linearization. Maybe some compliance detection as the following command reports the warning too:
$ qpdf --check h1.pdf

**Comment by jberkenbilt** [comment]
What qpdf does to check linearization is to calculate all the values of all the linearization hint tables and then compare with what's there. When I originally wrote that code, it was before qpdf was open source, and I was working for a company that processed thousands of PDF files a day. I found two things to be true: 1) a great many linearized PDF files have incorrect hint tables, and 2) most browser plugins actually use only a very small amount of the linearization data to present the files incrementally. The standard test people appear to use for whether linearization works correctly is whether or not the file starts to display in the browser before it is completely downloaded.

There are two conditions under which qpdf would issue a warning: an error in qpdf's code such that it incorrectly computes the linearization data and reports a problem when it differs from the input even if the input is correct, or the input is actually wrong. The linearization code in qpdf is very old and has had less attention than really any other part of the library. For 8.3.0, I did a pass through to make sure that problems always resulted in errors or warnings rather than abnormal termination, but it would still be useful to take pass through and make sure the code is actually correct. It would probably just take me a few hours to reread that part of the spec and take a pass through the code to see whether it's right. Most likely when I wrote that code, the latest PDF version was 1.4. If I were to check it over, I'd work with the ISO-32000 PDF 2.0 spec, which has clarified language in a lot of places where the original spec was ambiguous. (The PDF spec is still pretty hard to read in spots...a lot of stuff is left to the imagination of the reader.) The other thing is that the linearization code reports two different kinds of errors: places where the contents of the hint tables aren't as expected, and places where it finds inconsistencies in its own calculations. The latter case probably always indicates a qpdf bug. Unfortunately it's not obvious going through the code in all cases which type of error a given error is. This is all because I originally wrote that code as a checker for my own linearization write code.

That's probably too much information. Anyway, issues like this come up from time to time, so I think I will go ahead and prioritize taking a pass through the linearization code. A good start would be to provide a human-readable dump of the linearization tables. There is a `--show-linearization` option, but it's picky and won't show data if it finds any errors. This makes it very hard to debug this kind of problem. Almost surely fixable in about a day's work. It's just finding a day........

May I include your sample file in qpdf's test suite if I get to this?

**Comment by foxmadr** [comment]
Thank you very much. Yes, you can use my PDF as you need. My PDF is not linearized. I originally wanted to use your tool to linearize it.

**Comment by jberkenbilt** [comment]
Oh, I misunderstood. Yes, I see qpdf reports a problem when attempting to linearize. This is definitely a bug. I will look into it and try to get a fix for 8.4.0 which I am hoping to get out this month or no later than early February.

**Comment by jberkenbilt** [comment]
Thanks for following up and preventing me from overlooking a bug!

**Comment by jberkenbilt** [comment]
Actually, I see what's going on...the input file is actually malformed. qpdf is reporting this problem but still successfully learnizes the file. The file trailer reports the number of objects at 18 instead of 17. As you say, object 3 is overridden. qpdf is correct to count it once. This is a replacement of the original object. So it's not a bug. I need to slow down. :-) It is harmless though. This a place where the spec's wording is a bit ambiguous, but examples and clarifications in the spec indicate that you are not supposed to double-count replaced objects.

So this doesn't have anything to do with linearization at all.

Bottom line: the file has a recoverable problem that qpdf is reporting, and linearization works.

Should we go ahead and close? Sorry for the prolonged discussion over my just misunderstanding the issue.

**Comment by jberkenbilt** [comment]
I've studied the file and confirmed that the file is actually bad, the linearized file is correct in spite of the warning, and qpdf is behaving as intended. Sorry for the runaround. I'll go ahead and close this issue, but as always, please feel free to comment or reopen if needed.

**Comment by foxmadr** [comment]
Please what is bad in the PDF? I checked it by many online tools and PDF viewers and there was no problem reported.

**Comment by jberkenbilt** [comment]
The /Size key in the trailer dictionary is wrong. If you open this in a text editor, you can follow it. At the end of the file, the startxref key shows the trailer to start at offset 33931. This is an appended xref table that overrides object 3, which it does correctly. However, it sets /Size to 18, when the correct value is 17. The PDF spec is worded in a somewhat confusion fashion regarding /Size, but it is clarified in several spots. The PDF spec defines /Size as follows: "The total number of entries in the file’s cross-reference table, as defined by the combination of the original section and all update sections. Equivalently, this value shall be 1 greater than the highest object number defined in the file." One might thing this means that you have to count object 3 twice, but that's not what it means. The "Equivalently..." clause clarifies that it should be 1 greater than the highest object. The highest object in the file is object 16, and the document's original xref table has 17 entries (0 to 16) and is followed by the original trailer dictionary, which has /Size 17. In addition to the clarifying statement in the spec, there are also several examples of updating files by append in the appendix and in the text, and they do not add to /Size when updating an object. Also, you can find many applications that update PDF files by append, such as when saving form fields, and they do not add to /Size. The warning issued by qpdf: "WARNING: h1.pdf: reported number of objects (18) inconsistent with actual number of objects (17)" is reporting the discrepancy between /Size ("reported number of objects") and the actual number of objects in the xref table. Most validation tools probably just disregard this. There is really no use for the /Size entry as the xref table fully includes enough information to know this. More likely than not, if I were coding qpdf knowing what I know now about erroneous PDF files in the wild, I would simply ignore /Size, but since qpdf started its life as a PDF checker, it contains a lot of validations for which there is an unambiguous recovery.

So the one-sentence answer is that /Size should have the value 17 in the appended section of the PDF file. Hope that helps.
